common/src/main/java/com/wynntils/features/user/QuickCastFeature.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.user;
+
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.core.features.properties.FeatureInfo;
+import com.wynntils.core.features.properties.RegisterKeyBind;
+import com.wynntils.core.keybinds.KeyHolder;
+import com.wynntils.mc.event.ClientTickEvent;
+import com.wynntils.mc.utils.ItemUtils;
+import com.wynntils.mc.utils.McUtils;
+import com.wynntils.wc.event.WorldStateEvent;
+import com.wynntils.wc.utils.SpellManager;
+import com.wynntils.wc.utils.WynnItemMatchers;
+import com.wynntils.wc.utils.WynnUtils;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.ServerboundSetCarriedItemPacket;
+import net.minecraft.network.protocol.game.ServerboundSwingPacket;
+import net.minecraft.network.protocol.game.ServerboundUseItemPacket;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import org.lwjgl.glfw.GLFW;
+
+@FeatureInfo(stability = FeatureInfo.Stability.EXPERIMENTAL)
+public class QuickCastFeature extends UserFeature {
+
+    @RegisterKeyBind
+    private final KeyHolder castFirstSpell =
+            new KeyHolder("Cast 1st Spell", GLFW.GLFW_KEY_Z, "Wynntils", true, QuickCastFeature::castFirstSpell);
+
+    @RegisterKeyBind
+    private final KeyHolder castSecondSpell =
+            new KeyHolder("Cast 2nd Spell", GLFW.GLFW_KEY_X, "Wynntils", true, QuickCastFeature::castSecondSpell);
+
+    @RegisterKeyBind
+    private final KeyHolder castThirdSpell =
+            new KeyHolder("Cast 3rd Spell", GLFW.GLFW_KEY_C, "Wynntils", true, QuickCastFeature::castThirdSpell);
+
+    @RegisterKeyBind
+    private final KeyHolder castFourthSpell =
+            new KeyHolder("Cast 4th Spell", GLFW.GLFW_KEY_V, "Wynntils", true, QuickCastFeature::castFourthSpell);
+
+    private static final Pattern INCORRECT_CLASS_PATTERN =
+            Pattern.compile("§✖§ Class Req: (.+)".replace("§", "(?:§[0-9a-fklmnor])*"));
+    private static final Pattern LVL_MIN_NOT_REACHED_PATTERN =
+            Pattern.compile("§✖§ (.+) Min: ([0-9]+)".replace("§", "(?:§[0-9a-fklmnor])*"));
+    private static final Packet<?> RIGHT_CLICK_PACKET = new ServerboundUseItemPacket(InteractionHand.MAIN_HAND);
+    private static final Packet<?> LEFT_CLICK_PACKET = new ServerboundSwingPacket(InteractionHand.MAIN_HAND);
+    private static final Queue<Packet<?>> SPELL_PACKET_QUEUE = new LinkedList<>();
+    private static final boolean RIGHT = true;
```

Comment:
This also ties in with my comment about an enum in SpellManager.

Reply:
I ended up leaving them as boolean as I didn't want to make another enum for this.

Reply:
Can't you use the same enum for both? I'm thinking the only enum you need is PRIMARY/SECONDARY, all the way through, except at the very end, where you decide if primary is SwingPacket or UseItemPacket, depending on if you are an Archer or not.

Reply:
The main problem is that the ingame spell gui doesn't distinguish between an L as an archer or an L as a mage, the parser is gonna think see the same, even though one is primary and one is secondary. I could do the class detection while parsing the spell, but there it is not guaranteed that the player is still holding the weapon. Since auto-class-join class detection still isn't implemented Character can't be used to fetch the class reliably.

Reply:
Hm. Okay, let's keep it at this for the time being. I hope we can return to this later, if we have proper class detection implemented.

Or maybe we should not consider like "LRL" and "RLR" to be the "same" spell, and actually try to keep L and R properly separated. 

I dunno. I'm not 100% happy with the current solution but I don't have anything better to offer.